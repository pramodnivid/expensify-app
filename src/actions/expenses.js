import {createStore,combineReducers} from "redux"
import uuid from "uuid";

//multiple smaller functions
export const addExpense = (
                    {description='',
                    notes='',
                    amount=0,
                    createdAt = 0
                    } = {})=>{
                        console.log(description)
                       return {
    type:"ADD_EXPENSE",
    expense: {
        id: uuid(),description,notes,amount,createdAt
    }



}};



export const editExpense = (id,updates)=> ({

    type: 'EDIT_EXPENSE',
    id:id,
    updates

})

export const  removeExpense = ( {id}= {} )=>({
            type:"REMOVE_EXPENSE",
            id
                    
            });

