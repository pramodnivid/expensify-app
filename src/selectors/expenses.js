import moment from "moment"

const getVisibleExpenses = (expenses,{text,sortBy,startDate,endDate}) => {

    return expenses.filter((expense)=> 
    {  
       const createdAtMoment = moment(expense.createdAt)
        // const startDateMatch = typeof startDate !== 'number' || startDate<=expense.createdAt
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day') :true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') :true;

        //const endDateMatch = typeof endDate !== 'number' || endDate>=expense.createdAt
        const textMacth = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMacth;
    
    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt?1:-1;
        }

        else if(sortBy === 'amount'){
            return a.amount < b.amount?1:-1;
        }

    })



}
export default getVisibleExpenses;