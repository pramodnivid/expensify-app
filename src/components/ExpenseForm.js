import React from "react"
import {addExpense} from "../actions/expenses"
import {connect} from "react-redux"
import moment from "moment";
import {SingleDatePicker} from "react-dates"
const now = moment();
console.log(now.format('DD/MMM/YYYY'))



export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            description : props.expense ? props.expense.description: '',
            amount:props.expense?(props.expense.amount/100).toString():'',
            note : props.note ? props.expense.note:'',
            createdAt: props.expense ? moment(props.expense.createdAt):moment(),
            calendarFocused: false,
            error:undefined
    
        }
    
    }

    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>({description}))
    }

    onAmountChange = (e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }

    

    onNoteChange = (e)=>{
        const note = e.target.value;
        this.setState(()=>({note}))
    }
    addexp = (e) => {
        this.props.dispatch(addExpense({description:this.description}))
    }


    onDateChange = (createdAt) => {
        if(createdAt)
            this.setState(()=>({createdAt}))


    }

    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused:focused}))


    }

    onSubmit = (e)=>{
        e.preventDefault()
        if(!this.state.description | !this.state.amount ){
            this.setState(()=>({error:"Please enter amount and description"}))

        }else{
            this.setState(()=>({error:undefined}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) *100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })

            console.log("submit")
        }
    }

    render(){
        
        return (
        <div>
        <form onSubmit = {this.onSubmit}>
            <input type="text" name="description" 
                    value={this.state.description} 
                    onChange = {this.onDescriptionChange}
                    placeholder="Description" autoFocus>
            </input>
            
            <input type="text" name="amount" 
                    value={this.state.amount} 
                    onChange = {this.onAmountChange}
                    placeholder="Amount">

            </input>
            <SingleDatePicker
                date={this.state.createdAt}
                onDateChange = {this.onDateChange}
                onFocusChange = {this.onFocusChange} 
                focused = {this.state.calendarFocused}
                numberOfMonths = {1}
                isOutsideRange = {()=> false}        
            />
            <textarea 
            value={this.state.note} 
            onChange = {this.onNoteChange}

            placeholder="Add a note your expense"> 
            </textarea>
            <button>Add</button>
        </form>
        {!!this.state.error && <p> {this.state.error} </p>}
        </div>)
    }
}



//export default connect()(ExpenseForm)