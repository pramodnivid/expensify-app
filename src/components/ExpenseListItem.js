import React from "react"
import ReactDOM from "react-dom"
import {removeExpense} from "../actions/expenses"
import {connect} from 'react-redux'
import {BrowserRouter,Route,Switch,NavLink} from "react-router-dom"

export const ExpenseListItem = (props) =>(
    <div> 
        <h3>{props.description}</h3>
        <p>amount:{props.amount}, created on: {props.createdAt}</p> 
        <button onClick = { (e)=>{
            props.dispatch(removeExpense(props))
        }} > remove </button>
        <NavLink to={"/edit/"+props.id} activeClassName="is-active" exact={true}> Edit </NavLink>


    </div>
)

// const mapStateToProps = (state) => {
//     return {
//         filters: state.filters
//     }
// }
export default connect()(ExpenseListItem);
