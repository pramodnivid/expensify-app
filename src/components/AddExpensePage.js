import React from "react"
import ReactDOM from "react-dom"
import ExpenseForm from "./ExpenseForm"
import {addExpense} from "../actions/expenses"
import {connect} from "react-redux"
export class AddExpensePage extends React.Component{

    onSubmit = (expense)=>{
        console.log(expense)
        this.props.addExpense(expense);
        this.props.history.push('/');

    }

render(){
    return(
        <div>
        <h3>Add Expense</h3>
            <ExpenseForm onSubmit = {this.onSubmit}/>
        </div>
    )
}
}
const mapDispatchToProps = (dispatch) =>{
    return{
        addExpense: (expense)=>{
            dispatch(addExpense(expense))
        }
    };
}

export default connect(undefined,mapDispatchToProps)(AddExpensePage);
