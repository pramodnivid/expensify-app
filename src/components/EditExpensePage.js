import React from "react"
import ReactDOM from "react-dom"
import {editExpense,removeExpense} from "../actions/expenses"
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"

export class EditExpensePage extends React.Component{

    onSubmit = (expense)=>{
        //this.props.dispatch(editExpense(props.expense.id,expense))
        this.props.editExpense(this.props.expense.id,expense);

        console.log("upd",expense)
        this.props.history.push('/')

    }
    onClick = () => {
        
            this.props.removeExpense( {id: this.props.expense.id})
            this.props.history.push('/')
        

    }



render(){
    return (<div>
        <ExpenseForm expense={this.props.expense} onSubmit = {this.onSubmit}/>

        <button onClick = {this.onClick} > remove </button>

        </div>)

    }

}
const mapToStateProps = (state,props) =>{
    return {
        expense: state.expenses.find((expense) => {
            return (expense.id == props.match.params.id)
        }) 
    }

}

const mapDispatchToProps = (dispatch) =>{
    return{
        editExpense: (id,updates)=>{
            dispatch(editExpense(id,updates))
        },
        removeExpense: (id)=>{
            dispatch(removeExpense(id))
        }

    };
}


export default connect(mapToStateProps,mapDispatchToProps)(EditExpensePage);