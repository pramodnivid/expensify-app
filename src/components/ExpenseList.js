import React from "react"
import ReactDOM from "react-dom"
import {BrowserRouter,Route,Switch,NavLink} from "react-router-dom"
import { Provider } from "react-redux"

import {connect} from 'react-redux'
import expenses from "../reducers/expenses";
import ExpenseListItem from "./ExpenseListItem"
import selectExpenses from "../selectors/expenses"

export const ExpenseList = (props) =>(
    <div>
        <h1> List </h1>
        {console.log(props.expenses)}
        {props.expenses.length === 0 ? (<p> No expenses </p>) :
            props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense}/> )}

    </div>
)


const mapStateProps  = (state)=>{
    return {
        expenses: selectExpenses(state.expenses,state.filters),
        filters: state.filters

    };
}


export default connect(mapStateProps)(ExpenseList);


