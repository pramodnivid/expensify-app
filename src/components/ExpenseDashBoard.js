import React from "react"
import ReactDOM from "react-dom"
import ExpenseList from "./ExpenseList.js"
import ExpenseListFilters from "./ExpenseListFilters"
const ExpenseDashBoardPage=()=>(
    <div>
        <ExpenseListFilters/>
        <ExpenseList/>
    
    </div>
)
export default ExpenseDashBoardPage;