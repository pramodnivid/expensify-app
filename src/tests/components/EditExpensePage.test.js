

import React from "react"
import {shallow} from "enzyme"
import moment from "moment"
import {EditExpensePage} from "../../components/EditExpensePage"
import expenses from "../fixtures/expenses"

let editExpense,removeExpense,history,wrapper;
beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    history = {push: jest.fn()}
    wrapper = shallow(<EditExpensePage editExpense = {editExpense} removeExpense = {removeExpense} expense = {expenses[0]} history={history} />)

})

test("should render edit page correctly ",()=>{
    expect(wrapper).toMatchSnapshot();
})

test("should handle edit onSubmit",()=>{
     wrapper.find('ExpenseForm').prop('onSubmit')({amount:123});
     expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,{amount:123});

 })

 test("should handle remove onSubmit",()=>{
    wrapper.find('button').prop('onClick')('1');
    expect(history.push).toHaveBeenLastCalledWith('/');
   expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[0].id});

})
