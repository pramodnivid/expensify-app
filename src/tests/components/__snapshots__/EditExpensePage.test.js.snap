// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit page correctly  1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 132,
        "createdAt": 0,
        "description": "rent",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     remove 
  </button>
</div>
`;
